### ISSUE TITLE
Add layer-wise adaptive rate scaling (LARS) optimizer

### ISSUE THREAD

**Issue by noahgolmant** [pytorch/pytorch#18414]
## 🚀 Feature

Add a PyTorch implementation of layer-wise adaptive rate scaling (LARS) from the paper "[Large Batch Training of Convolutional Networks](https://arxiv.org/abs/1708.03888)" by You, Gitman, and Ginsburg. Namely, include the implementation found [here](https://github.com/noahgolmant/pytorch-lars) in the `torch.optim.optimizers` submodule.

## Motivation

LARS is one of the most popular optimization algorithms for large-batch training. This feature will expose the algorithm through the high-level optimizer interface. There are currently no other implementations of this algorithm available in widely used frameworks.

## Pitch

I have already written/tested a PyTorch optimizer for LARS in [this repo](https://github.com/noahgolmant/pytorch-lars). Changes would be minimal (I just have to make the imports relative at the top of the file).

## Alternatives

The only existing code for this is in caffe, and it looks like it is not exposed at a higher level within the solver framework. I also do not know whether or not that code has been tested.

## Additional context

Notably, I was unable to reproduce their results (namely reducing the generalization gap) on CIFAR-10 with ResNet. I would welcome additional tests to see if this implementation is able to replicate their performance on larger datasets like ImageNet.

**Comment by fmassa** [comment]
cc @soumith what do you think?

**Comment by soumith** [comment]
I'm supportive of adding LARS, it's been used in large-scale learning for a while now.

cc: @thatguymike

**Comment by ngimel** [comment]
Apex has LARC implementation https://github.com/NVIDIA/apex/blob/d74fda260c403f775817470d87f810f816f3d615/apex/parallel/LARC.py

**Comment by noahgolmant** [comment]
I wasn't aware of the existing Apex implementation, thank you for pointing it out (searching for LARS implementations did not lead to any results for optimizers named LARC). Either way, I am supportive of adding the optimizer upstream.

The existing Apex implementation serves as a wrapper to other optimizers. I do not know of any experiments where LARS/LARC was used to modify optimizers other than SGD with momentum, but either way I think it would make more sense to include it as its own optimizer if we were to add it to the `optimizers` submodule.

**Comment by slowbull** [comment]
There are two minor issues in your code which may cause your bad results on BS 16384. I can obtain test error 85% on BS16384 with my implementation.    Firstly, You don't need to apply lars on 'bn' and 'bias'.  Adding these lines may help.  

params_skip = []                                                                                                                                                                             
  params_noskip = []                                                                                                                                                                           
  skip_lists = ['bn', 'bias']                                                                                                                                                                  
  for name, param in net.named_parameters():                                                                                                                                                   
    if any(name in skip_name for skip_name in skip_lists):                                                                                                                                     
      params_skip.append(param)                                                                                                                                                                
    else:                                                                                                                                                                                      
      params_noskip.append(param)                                                                                                                                                              
  param_lrs = [{'params':params_skip, 'lr':state['learning_rate']},                                                                                                                            
                {'params':params_noskip, 'lr':state['learning_rate']}]                                                                                                                         
 
The other one is that taking care of using chunk_size. Do make sure that optimizer.step() is processed only after accumulating enough gradients. So you need a variable to count it in the lar.py file.

**Comment by kamathhrishi** [comment]
@noahgolmant , any progress?

**Comment by AtlantixJJ** [comment]
Hi, I need LARS in Pytorch to conduct a research. Do you have any progress? I can help you if you need it.

**Comment by sublee** [comment]
Hi, I'm Heungsub Lee at Kakao Brain. I'm interested in large-scale training using PyTorch. As this aspect, I've implemented GPipe in PyTorch as [torchgpipe](https://github.com/kakaobrain/torchgpipe).

If still there's no one who is implementing accelerated LARS in PyTorch yet, I want to try it. May I give a chance to contribute to PyTorch?

**Comment by soumith** [comment]
@sublee no one is implementing LARS in PyTorch yet, go ahead.
torchgpipe is a great project btw

**Comment by sublee** [comment]
@soumith Thank you for letting me know. My co-worker is currently involved in a PyTorch LARS project optimized for CUDA. It would be released as a separate project. Anyways, we may port it into PyTorch itself later.

**Comment by hanyc0914** [comment]
> There are two minor issues in your code which may cause your bad results on BS 16384. I can obtain test error 85% on BS16384 with my implementation. Firstly, You don't need to apply lars on 'bn' and 'bias'. Adding these lines may help.
> 
> params_skip = []
> params_noskip = []
> skip_lists = ['bn', 'bias']
> for name, param in net.named_parameters():
> if any(name in skip_name for skip_name in skip_lists):
> params_skip.append(param)
> else:
> params_noskip.append(param)
> param_lrs = [{'params':params_skip, 'lr':state['learning_rate']},
> {'params':params_noskip, 'lr':state['learning_rate']}]
> 
> The other one is that taking care of using chunk_size. Do make sure that optimizer.step() is processed only after accumulating enough gradients. So you need a variable to count it in the lar.py file.



> There are two minor issues in your code which may cause your bad results on BS 16384. I can obtain test error 85% on BS16384 with my implementation. Firstly, You don't need to apply lars on 'bn' and 'bias'. Adding these lines may help.
> 
> params_skip = []
> params_noskip = []
> skip_lists = ['bn', 'bias']
> for name, param in net.named_parameters():
> if any(name in skip_name for skip_name in skip_lists):
> params_skip.append(param)
> else:
> params_noskip.append(param)
> param_lrs = [{'params':params_skip, 'lr':state['learning_rate']},
> {'params':params_noskip, 'lr':state['learning_rate']}]
> 
> The other one is that taking care of using chunk_size. Do make sure that optimizer.step() is processed only after accumulating enough gradients. So you need a variable to count it in the lar.py file.

Hi, I wonder why bn ans bias doesn't need apply lars optimizer, because I couldn't find any explaination in paper. Is there any proof or official elaboration? Thank you very much!

**Comment by sytelus** [comment]
Hi @sublee - just wondering if there was any update on LARS/LARC implementation... Thanks!

**Comment by sublee** [comment]
Hi @sytelus, Thank you for asking. My co-worker released the project named `torchlars`: https://github.com/kakaobrain/torchlars

**Comment by sytelus** [comment]
That looks great! Thanks so much for releasing this!!

**Comment by guptv93** [comment]
Does this work with nn.DataParallel? I keep getting segmentation fault when I try to use this with a DataParallel model.

**Comment by turian** [comment]
I am still interested in LARS. It is used in VicReg (https://arxiv.org/pdf/2105.04906.pdf) and BYOL and Barlow Twins self-supervised learning.

Here is another implementation of it:
https://github.com/NUS-HPC-AI-Lab/LARS-ImageNet-PyTorch

And the vicreg implementations seems to have a simple pure torch implementation that works distributed too:
https://github.com/facebookresearch/vicreg/blob/main/main_vicreg.py#L245
